### PR TITLE
busybox: enable telnet on normal devices

### PR DESCRIPTION
**Pull request by aparcar**
There is no telnet compatible tool available for OpenWrt ever since it
was removed with the switch to OpenSSH for logins.

This commits adds telnet if the device is not a small flash one.

Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

**Comment by pkgadd** [comment]
IMHO it feels wrong to enable this in busybox. Admittedly, it's the easiest approach, but it adds 'useless' (no one has missed telnet support in almost a decade) weight to the most essential/ monolithic package in OpenWrt and it doesn't provide any means for the smallflash devices to install it.

Wouldn't it make more sense to package a small dedicated telnet client (BSD or similar) as an independent package? Yes, that will be larger than enabling the telnet-client feature in busybox, but it's an optional package barely anyone will need to install.

**Comment by pkgadd** [comment]
E.g. merging this https://github.com/openwrt/packages/pull/4509

**Comment by mrkiko** [comment]
I guessit this depends on your current needs.
f you need full telnet and ANSI support, then you may need a telnet client. Otherwise, the "nc" utility might be enough.

**Comment by aparcar** [comment]
@mrkiko in my case it sadly doesn't work :)

@pkgadd thanks for pointing me to that PR. Having telnet in busybox would reduce maintenance however if someone is fine with maintaining telnet, great!

**Comment by f00b4r0** [comment]
> IMHO it feels wrong to enable this in busybox. Admittedly, it's the easiest approach, but it adds 'useless' (no one has missed telnet support in almost a decade) weight to the most essential/ monolithic package in OpenWrt and it doesn't provide any means for the smallflash devices to install it.

IMHO the latter argument is rather moot: currently *no device* can install a telnet client anyway. Adding such a client while leaving SMALL_FLASH out of the fray already seems like an *improvement* over the current situation.
Furthermore SMALL_FLASH is also so utterly constrained that running current master on these devices is most likely to require a manual, tightly crafted build, at which point the knowledgeable user can toggle this setting as they see fit.

> Wouldn't it make more sense to package a small dedicated telnet client (BSD or similar) as an independent package? Yes, that will be larger than enabling the telnet-client feature in busybox, but it's an optional package barely anyone will need to install.

The problem likely is, as shown in the suggested https://github.com/openwrt/packages/pull/4509, that nobody wants to maintain that. Enabling is busybox is a low hanging fruit.

As for the needs for a telnet client, I've experienced first hand how `nc` can be very insufficient at the task to connect to some devices which do not offer an ssh alternative, even though connecting to these devices from the OpenWrt-running one may be desirable :)

My 2c.

**Comment by tapper82** [comment]
Is telnet  safe? Should it be disabled by default. Would this be exposed in luci so we can toggle it on or off.

**Comment by pkgadd** [comment]
I totally understand the desire of having a telnet client available (beyond of what nc has to offer), my only issue with this is adding it to busybox, as this bloats an essential and rather tricky package (there is a reason why the buildsystem marks private builds with modified busybox settings as tainted) for a fringe benefit. Yes, this bloat might only be 8.8 KB (as indicated by @aparcar on IRC), but those are 8.8 KB for a feature that isn't needed for hardware support, nor the OpenWrt userland at all - it's a luxury item that makes using OpenWrt as jumphost, VM/ CI target or similar more convenient. While there are requests for telnet support coming in somewhat regularly (50:50 on client and server requests, the later probably for commercial endeavours to tick it off the requirement list), still only a handful of users will ever want to have it on their router. This makes this imho more of a case for separate package, than bloating the base install (this would be different if the feature could be split out to a non-default binary package, but that's not really an option for busybox (apart from providing some kind of "busybox-extra" package shipping an optional full(er)-featured busybox (let's say `/bin/busybox-extra`) in addition to the `/bin/busybox` and letting that fullfil those extra symlinks)). Keep in mind that quite a few devices are rather tight on flash, e.g. think of the 6976 KB max. for the ZyXEL gs1900 series (and there are quite a few other devices with less than ideal vendor partitioning without much breathing space left). I don't think size constraints (we're talking about <<100 KB (telnet-bsd, I didn't check the exact size, yes, but it should be well under that size) to maybe 400 KB extra for a second busybox-extra package would be a problem for the kind of system where a telnet client might be desired, but needlessly 'wasting' 10 KB on a fringe feature in the most core part of the system eats away the (already tight) margins without a reason.

Sure, from a maintenance point of view, just enabling it in busybox and then forgetting about it is easier - but if this feature is so valuable, maintaining telnet-bsd shouldn't be much of a hindrance (the package is small, simple, rather stable - and, thanks to diizzyy, 99.8% of the packaging is already done - all it needs is a maintainer and (likely) applying those two additional patches, which allow dropping `-Wno-format-security`, and maybe hardcoding an upstream location instead of using `@DISTFILES`). I'd even volunteer to do those last 0.2% of polishing, but I won't stand up as maintainer - because, frankly, I don't need a telnet client on my router.

**Comment by pkgadd** [comment]
@tapper82 this PR has been about adding a telnet client, not about a telnet**d** (server), it would be used interactively from OpenWrt's shell to access other systems (which do run a telnet-server, e.g. lots of commercial networking gear, like managed enterprise switches, etc.) or to debug various other services (e.g. I have used it to debug POP3/ SMTP servers in the past and even occasionally read my mail with a telnet client (no, it's not fun, but there may be reasons)).

So while telnet is not secure at all, there is no security risk of providing a telnet **client** - after all it's totally inactive until you explicitly start using it (from the ssh shell, by invoking `telnet [host [port]]`, it's 'just' a problem of needlessly bloating the default firmware size for rather exotic use cases.

**Comment by pkgadd** [comment]
Just for reference (x86_64):
```
$ ls -gG ./bin/packages/x86_64/packages/telnet-bsd_1.2-1_x86_64.ipk
-rw-r--r-- 1 36495 17. Aug 01:48 ./bin/packages/x86_64/packages/telnet-bsd_1.2-1_x86_64.ipk

$ tar -Oxf ./bin/packages/x86_64/packages/telnet-bsd_1.2-1_x86_64.ipk ./data.tar.gz | tar -tzvf -
drwxr-xr-x 0/0               0 2022-08-17 00:51 ./
drwxr-xr-x 0/0               0 2022-08-17 00:51 ./usr/
drwxr-xr-x 0/0               0 2022-08-17 00:51 ./usr/bin/
-rwxr-xr-x 0/0           84475 2022-08-17 00:51 ./usr/bin/telnet
```

**Comment by f00b4r0** [comment]
Just to give some perspective here:
- Looking at the mt7621 folder on d.o.o, the average growth of all sysupgrade images was **22.2KB** from 22.03.0-rc1 to 22.03.0-rc6
- The extra "8.8 kb" quoted by @aparcar earlier is the menuconfig-indicated *binary size* increase and rather irrelevant in the context of a compressed filesystem. As it turns out, the figure itself might be quite overblown: I test built a mt7621 Yuncore AX820 image with/without `CONFIG_BUSYBOX_CONFIG_TELNET`, the resulting `busybox` binary was only **1124 bytes** bigger, and the final squashfs image size was **identical**, suggesting that the extra "bloat" is entirely offset by the squashfs compression/overhead here.
- FWIW `mkswap`, `swapon` and `swapoff` are unconditionally enabled on all devices, allegedly representing an extra "35 kb" binary size (according to the busybox menuconfig): I suppose one could also argue of their necessity on most embedded devices. Incidentally, disabling them *also* didn't change the final image size for this device (the busybox binary was only 628 bytes smaller, again suggesting the menuconfig figures cannot be trusted).

Maybe we can spare ourselves a heated debate in the face of such insignificant numbers? 😉 

@pkgadd I'm sure you can see that your argument can be entirely flipped over, whereby enabling a telnet client in busybox at next to no size cost (as evidenced above) can provide this *feature* to a lot of devices that would otherwise be unable to install a package that's several orders of magnitude bigger.

HTH

**Comment by aparcar** [comment]
As a side comment regarding those benchmarks, from my understanding the squashfs is padded so you should compare the rootfs.tar.gz instead.

I'll put a bit more thought into this, my use case is migrating LAVA to OpenWrt which sadly requires a full blown telnet client, another approach could be to fix this in LAVA instead.

**Comment by f00b4r0** [comment]
@aparcar unless I'm mistaken there is no rootfs.tar.gz involved for squashfs devices: squashfs has its own file structure and compression (often LZMA).
As for your second point, it doesn't preclude the benefit of having a telnet client on devices for other purposes (such as those highlighted [above](https://github.com/openwrt/openwrt/pull/10462#issuecomment-1217244378).

**Comment by aparcar** [comment]
This is superseded by https://github.com/openwrt/packages/pull/19191 

I really don't want to bloat the system and I'd be happier not so rely on `ser2net` and `telnet` for my use case. However until other things change, I decided `telnet-bsd` is the way to go. Enabling an insecure and outdated protocol per default (even if only client) on devices doesn't seem to be the way forward.

**Comment by f00b4r0** [comment]
@aparcar i'm sorry that you seemed to have completely ignored my arguments.

**Comment by aparcar** [comment]
> Enabling an insecure and outdated protocol per default (even if only client) on devices doesn't seem to be the way forward.

We could enable so many more small things which are small but for what gain? Do **you** have a need for `telnet`? Specifically `telnet` as part of `busybox` instead of an external binary?

**Comment by f00b4r0** [comment]
> > Enabling an insecure and outdated protocol per default (even if only client) on devices doesn't seem to be the way forward.
> 
> We could enable so many more small things which are small but for what gain? Do **you** have a need for `telnet`? Specifically `telnet` as part of `busybox` instead of an external binary?

I certainly do (to access e.g. some iot devices and switches): I was hit by the lack of telnet client before and had to go through the pain of nc which can only do so much. Busybox telnet would do just fine, I don't need to actually waste an extra 80KB for a standalone client. The "insecure" argument is moot: using nc is equally insecure in that context. We're talking about a *client*: I'm sure you're not suggesting we should remove http support from the wget applet because it's insecure?

Anyway, my point was that enabling this extra applet clearly doesn't bloat anything and thus adds "at (almost) no cost" an extra feature to the image: imho this is a win over a full blown telnet client that *is* measurably larger.

HTH

**Comment by lucize** [comment]
oh, how many times I wished a telnet client to be available by default ! hope this goes in..

**Comment by the2masters** [comment]
Why do you need a Telnet client on your router? You log into the router with ssh and then telnet into another device? You could telnet directly into the device from your laptop. If it's remote you could setup a ssh port forwarding:
`ssh -L 23:192.168.0.111:23 remoterouter.dyndns.org`

**Comment by lucize** [comment]
I guess workaround can always be fun, especially when you have to jump, then jump, then again jump and you have to connect to devices that don't have a default gw

**Comment by mpratt14** [comment]
I vaguely remember that very old openwrt had a feature where (I think) SSH was locked until a password was made with telnet or something like that. Or maybe it was the other way around...

I think telnet access should be locked until a root password is given. But I don't know how that was done...

**Comment by mpratt14** [comment]
the idea I'm referring to is the *reverse* of this, for reference (although it's so old that the diffs do not serve as examples)

https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=eae7eaa43db69af831eecf7e9ce0616f617564ee

**Comment by f00b4r0** [comment]
@mpratt14 this PR is not about enabling telnet **access to** devices; it is about enabling the telnet **client** on the device, so that a user that logs into their OpenWrt device can then **telnet to** another system.

Maybe the PR title/commit subject should be clarified, e.g. "busybox: enable telnet client" (the !SMALL_FLASH distinction is sufficient in the commit message imho).

**Comment by mpratt14** [comment]
title is fine, busybox distinguishes between client and server with the names `telnet` and `telnetd`, it just flew over my head until after I posted 😆

**Comment by Ansuel** [comment]
IMHO keeping telnet would create a security issue... online there are many guide that still say to use telnet to connect to modems/router. Shipping something like that would make someone use telnet instead of correct ssh connection... 

Considering an alternative solution is provided why should we include telnet in the default image if an user can just install a package and be done with it?

I understand some iot device still use telnet as a transmission medium (I myself have a xiaomi shit smart light that use that) but I honestly think we should at least try to prevent bad behaviours from the user.

**Comment by robimarko** [comment]
@Ansuel This is just the telnet client, not the server so you can connect to telnet only stuff.
Personally I had a need for telnet client in OpenWrt multiple times and its really annoying that there is none.

**Comment by f00b4r0** [comment]
@Ansuel please let's be careful with our choice of words: having the telnet client enabled by default is no more a security issue than having an http client (wget) enabled by default, or having netcat enabled by default. Not to mention that having the client and actually using it are two separate things.

I don't think we should infantilise users by deciding what's "good for them", especially when OpenWrt is about empowering users AIUI. Every tool can be misused in an insecure manner after all.

The key argument in favor of having telnet vs only nc is terminal emulation: using nc is a royal pain in that context.

As for the packaged alternative, it makes no "security" difference: whether packaged or provided by busybox it's always a telnet client "insecurely" used by the same set of users. However the packaged client is much bigger, potentially making it impossible to install on some devices, and it will require maintenance (whereas the busybox telnet client maintenance is no additional burden over busybox maintenance itself).

**Comment by Ansuel** [comment]
I understand this is only for client.

> I don't think we should infantilise users by deciding what's "good for them", especially when OpenWrt is about empowering users AIUI. Every tool can be misused in an insecure manner after all.

My concern is with how telnet was widely used back on the days and the fact that there are still may guide on the web that keep using them... An user would notice that telnet is present and would use that instead of checking if a better alternative is present. (with an external package an user would require extra step and would search a little more on how to accomplish his task)

But I understand that my argument is not that strong.

Anyway to sum up this:
- Size increase is 1kb more or less.
- BSD alternative is much bigger?
- Usecase: iot device that are controlled by script present on the router?
- Security concern is minimal as at the very end it's really user fault and using an external package would just adds another package to maintain.

All correct?

**Comment by f00b4r0** [comment]
@Ansuel all correct indeed :)
Size of the BSD alternative can be inferred here: https://github.com/openwrt/openwrt/pull/10462#issuecomment-1217295553

**Comment by f00b4r0** [comment]
Any chance that this PR could get out of "Draft" status and seriously considered? Alternatively it should probably be closed.

**Comment by Spudz76** [comment]
If you need telnet simply compile your own image, because telnet isn't the only thing missing, and the builtin `vi` features suck, and there is no LUCI by default, and there is no tcpdump, and... we could add "things that should be there for diagnostics" all day long.  Which I do, in my own builds, and then I'm happy.

In other words even if telnet were added by default there are still a hundred other reasons to just compile your own.

**Comment by mpratt14** [comment]
on one hand I agree but on the other hand, the OS is kind of expected to have all the basic network tools available. for example, I have thought of a way to get rid of `iw` to save space, using our own `iwinfo` but that probably wouldn't be acceptable because many devs expect `iw` to be in the OS like pretty much every other popular linux distro

**Comment by Spudz76** [comment]
Except this isn't really a distro and is not intended to be your workstation.  It's an embedded small OS.  You'd hate all the missing stuff in VxWorks haha.

Same reason there is no command history/undo in the busybox `vi` by default, which I hate and is stupid, but hey it saves space and I can just add `vim-fuller` and get every missing feature how I like it.  But really I should be using vim from by Debian-ish workstation and edit the files via sshfs or something instead of on-device, where space is limited and included junk should stay as light as still operates.  Of course it's hard to have all your favorite tools on a road trip but that cheap OEM lug wrench will work despite the existence of impact drivers and sockets, but you don't have space in  your car for all that weight AND a whole compressor and tank...

Think more like what you'd bring backpacking, which won't be the Keurig machine and a generator just so you can have your usual familiar easy coffee.

`netstat` doesn't come with a default install of Ubuntu Server.  Literally everything should have `netstat` but it doesn't and it's a real distro...

**Comment by aparcar** [comment]
I'm in favor of an optional package over at the community feeds instead of adding this by default. My use case is to special to have it enabled by default :)
